File: src/main/java/org/quizbe/model/Rating.java
package org.quizbe.model;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.Objects;

@Entity
@Table(name = "RATING",  uniqueConstraints={@UniqueConstraint(columnNames={"question_id", "user_id"})})
public class Rating {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "ID", nullable = false)
    private Long id;

    @Basic
    @Column(nullable = true)
    private String comment;

    @Basic
    @Column(name = "VALUERATING", nullable = false)
    private Integer value;

    @Basic
    @Column(nullable = false)
    private LocalDateTime dateUpdate;

    @ManyToOne
    private Question question;

    @ManyToOne
    private User user;

    public Rating(Long id, String comment, Integer value, LocalDateTime dateUpdate, Question question, User user) {
        this.id = id;
        this.comment = comment;
        this.value = value;
        this.dateUpdate = dateUpdate;
        this.question = question;
        this.user = user;
    }

    public Rating() {

    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public Integer getValue() {
        return value;
    }

    public void setValue(Integer value) {
        this.value = value;
    }

    public LocalDateTime getDateUpdate() {
        return dateUpdate;
    }

    public void setDateUpdate(LocalDateTime dateUpdate) {
        this.dateUpdate = dateUpdate;
    }

    public Question getQuestion() {
        return question;
    }

    public void setQuestion(Question question) {
        this.question = question;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public boolean isOutDated() {
        return question.getDateUpdate().isAfter(this.getDateUpdate());
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Rating rating = (Rating) o;
        return getId() == rating.getId() && Objects.equals(getComment(), rating.getComment()) && getValue().equals(rating.getValue()) && getDateUpdate().equals(rating.getDateUpdate());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getId(), getComment(), getValue(), getDateUpdate());
    }

    @Override
    public String toString() {
        return "Rating{" +
                "id=" + id +
                ", comment='" + comment + '\'' +
                ", value=" + value +
                ", dateUpdate=" + dateUpdate +
                '}';
    }
}

File: src/main/kotlin/org/quizbe/model/Rating.kt
package org.quizbe.model

import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "rating", uniqueConstraints = [UniqueConstraint(columnNames = ["question_id", "user_id"])])
class Rating {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    var id: Long? = null

    @Lob
    @Column(nullable = false)
    var comment: String = ""

    @Basic
    @Column(name = "valuerating", nullable = false)
    var value: Int? = null

    @Basic
    @Column(nullable = false)
    var dateUpdate: LocalDateTime? = null

    @ManyToOne
    var question: Question? = null

    @ManyToOne
    var user: User? = null

    constructor(id: Long?, comment: String?, value: Int?, dateUpdate: LocalDateTime?, question: Question?, user: User?, isObsolete: Boolean = false) {
        this.id = id
        this.comment = comment ?: ""
        this.value = value
        this.dateUpdate = dateUpdate
        this.question = question
        this.user = user
        this.isObsolete = isObsolete
    }

    constructor() {}

    val isOutDated: Boolean
        get() = question!!.dateUpdate!!.isAfter(dateUpdate)

    var isObsolete: Boolean = false

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val rating = o as Rating
        return id === rating.id && comment == rating.comment && value == rating.value && dateUpdate == rating.dateUpdate
    }

    override fun hashCode(): Int {
        return Objects.hash(id, comment, value, dateUpdate)
    }

    override fun toString(): String {
        return "Rating{" +
                "id=" + id +
                ", comment='" + comment + '\'' +
                ", value=" + value +
                ", dateUpdate=" + dateUpdate +
                '}'
    }
}
